File: 0198_House_robber/0198_house_robber.go

func max(num1, num2 int) int {
    if num1 > num2 {
        return num1
    }
    return num2
}

func rob(nums []int) int {
    n := len(nums)

    if n == 1 { 
        return nums[0]; 
    } else if n == 2 {
       return max(nums[0], nums[1])
    }

    prev := nums[0]
    curr := max(nums[0], nums[1])

    for i:=2; i<n; i++ {
        temp := curr
        curr = max((nums[i] + prev),curr)
        prev = temp
    }

    return curr;
}

File: 0198_House_robber/0198_house_robber.c
int rob(int* nums, int numsSize) {
    if(numsSize == 1) { return nums[0]; }
    else if(numsSize == 2) {
       return nums[0]>nums[1] ? nums[0] : nums[1];
    }

    int prev = nums[0];
    int curr = nums[0]>nums[1] ? nums[0] : nums[1];

    for(int i=2; i<numsSize; i++) {
        int temp = curr;
        curr = (nums[i] + prev) > curr ? nums[i] + prev : curr;
        prev = temp;
    }

    return curr;
}